sw/device/lib/base/memory.h

```diff
@@ -0,0 +1,85 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef SW_DEVICE_LIB_BASE_MEMORY_H_
+#define SW_DEVICE_LIB_BASE_MEMORY_H_
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdnoreturn.h>
+
+/**
+ * Load a word from memory directly, bypassing aliasing rules.
+ *
+ * ISO C forbids, in general, casting a pointer to non-character types and
+ * reading them, though it is frequently necessary to read exactly one word out
+ * of a |void *|. This function performs that action in a manner which is
+ * well-defined.
+ *
+ * Of course, |ptr| must point to word-aligned memory that is at least one word
+ * wide. To do otherwise is Undefined Behavior. Note that this memory need not
+ * be initialized.
+ *
+ * @param ptr a word-aligned pointer pointed to at least four bytes of memory.
+ * @return the word |ptr| points to.
+ */
+inline uint32_t read_32(void *ptr) {
+  uint32_t out;
+  asm volatile("lw %0, 0(%1)" : "=r"(out) : "r"(ptr));
```

Comment:
I don't think we need to do this, honestly.

You should be able to cast a `void*` to a `uint32_t*`, which is valid as long as the `void*` is aligned, which this function requires anyway. This has the advantage of totally avoiding inline (volatile) assembly.

The rules of aliasing don't really apply, as `void*` is assignment compatible with any other pointer type - C just makes it UB to dereference an underaligned pointer, a null pointer, or a few other kinds of pointer that don't apply in this case.

Reply:
I don't think that's true, unfortunately. My read of sections 6.5.5 and 6.5.6 is that if you have a `void*`, and you do not know a priori its effective type (per S6.5.5), the only lvalue you're guaranteed to know is ok is `char*`. In particular, my read of strict aliasing is that this is UB:
```
char bytes[4] alignas(4) = "abcd";
// ... across a function boundary ...
uint32_t word = *((uint32_t *)bytes);
```

Mind, normally I wouldn't care about function boundaries, but the explicit inline-ability here makes me nervous.